docs/proposals/config_data.md

```diff
@@ -0,0 +1,262 @@
+
+## Configuration Workflow
+
+The component configuration workflow is given below:
+
+1. ``ConfigData`` object is created through the API server.
+2. New components using the ``ConfigData`` object are created.
+3. ``ConfigData`` object is retrieved using the downward API.
+4. ``Data`` key/value pairs of ``ConfigData`` will be exposed as either environment variables (``EnvVarSource``) or as a new volume (``ConfigDataVolumeSource``).
+5. API server is watched for modifications to the object.
+6. After a modification, actions 3, 4, and 5 are repeated.
+
+**Note**: Environment variables specified in ``ConfigData`` will override the environment variables
+with the same name found in the system if consumed as environment variables.
+
+## Use Cases
+
+#### Pod Configuration
+
+```json
+{
+    "apiVersion": "v1",
+    "name": "PodConfiguration1",
+    "kind": "ConfigData",
+    "data": {
+        "API_SERVER": "10.240.13.14:8998",
+        "API_VERSION": "v1",
+        "DB_HOST": "10.240.55.3",
+        "DB_PORT": "3306"
+    }
+}
+```
+``PodConfiguration1`` is intended to be consumed as environment variable key/value pairs.
+
+```json
+{
+    "apiVersion": "v1",
+    "name": "PodConfiguration2",
+    "kind": "ConfigData",
+    "data": {
+        "config.yaml": "CONFIG_YAML_STRING_CONTENTS",
+        "spec.yaml": "SPEC_YAML_STRING_CONTENTS"
+    }
+}
+```
+``PodConfiguration2`` is intended to be used as a mounted volume containing two files with their respective contents.
+
+Then, configuration objects are persisted to the API server as follows:
+```bash
+kubectl create -f pod-configuration-1.json 
+kubectl create -f pod-configuration-2.json
+```
+
+We create a replication controller specification that creates pods using these two configuration objects:
+```yaml
+apiVersion: v1
+kind: ReplicationController
+metadata:
+  name: ExampleReplicationController
+spec:
+  replicas: 2
+  selector:
+    app: dummy-app
+  template:
+    metadata:
+      labels:
+        app: dummy-app
+    spec:
+      # New pods will be created using the ConfigData objects.
+      # Each config object will be updated be it a volume or environment variables.
+      # But, some of them might require a restart on update.
+      # Note: If multiple env ConfigData objects are specified, let's call them Env1 and Env2.
+      #       If both of them contains a same named environment variable, Env2 will be the final one.
+      #       If Env1 is updated, the reload process will be Env1 first, then Env2.
+      #       Even though Env2 is not updated, the 'env' ConfigData objects are reloaded in the
+      #       order they are specified to prevent Env1 from overriding any of the variables in Env2.
+      # TODO: 'config-data' is a separate field because there are different ways to consume the object
+      #       and 'volume' might be a bit too specific. More discussions on this will be needed.
+      config-data:
```

Comment:
Probably a prestart hook.

On Sep 14, 2015, at 7:31 PM, Andy Goldstein <notifications@github.com>
wrote:

In docs/proposals/config_data.md
<https://github.com/kubernetes/kubernetes/pull/6477#discussion_r39460019>:

> +  template:
> +    metadata:
> +      labels:
> +        app: dummy-app
> +    spec:
> +      # New pods will be created using the ConfigData objects.
> +      # Each config object will be updated be it a volume or environment variables.
> +      # But, some of them might require a restart on update.
> +      # Note: If multiple env ConfigData objects are specified, let's call them Env1 and Env2.
> +      #       If both of them contains a same named environment variable, Env2 will be the final one.
> +      #       If Env1 is updated, the reload process will be Env1 first, then Env2.
> +      #       Even though Env2 is not updated, the 'env' ConfigData objects are reloaded in the
> +      #       order they are specified to prevent Env1 from overriding any of the variables in Env2.
> +      # TODO: 'config-data' is a separate field because there are different ways to consume the object
> +      #       and 'volume' might be a bit too specific. More discussions on this will be needed.
> +      config-data:

But, the objects will contain multiple fields wouldn't that be
counter-productive?

Including the object in sometimes in volume section and sometimes in env
section seems unnecessary.
There also might be more ways to consume ConfigData or more options for the
ConfigData objects.

As Clayton wrote, the only 2 mechanisms in use today are env vars and
volumes. I really don't think we want a separate configData section in the
pod spec.

There *may* be some value in finding a way to easily pull in an entire set
of ConfigData key/value pairs as env vars without needed to explicitly and
manually list each one (and only if you didn't need any remapping). I'm not
sure how we'd do that exactly, however.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/6477/files#r39460019>.